### PR TITLE
Units, Parts, and Refits hold actual Person instances

### PR DESCRIPTION
**Pull request by sixlettervariables**
As with the part ID fix, this switches from person IDs littered throughout Unit, Part, and Refit.

There are two behavioral changes to this PR as well:
1. Importing parts no longer merges spares (we were already doing this later)
2. Spare parts do not include parts queued for refits or overnight work
    - I've added a test for this as well

Both of these caused issues with the ID fixups on "weird" campaigns.

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/io/CampaignXmlParser.java [hunk 1]
```diff
@@ -741,20 +729,20 @@ private void fixupUnitTechProblems(Campaign retVal) {
...
-                } else if (null == u.getTechId()) {
+                } else if (null == u.getTech()) {
                     reason = "was not referenced by unit";
                     u.setTech(tech);
                 } else if (u.isMothballed()) {
                     reason = "referenced mothballed unit";
                     unitDesc = u.getName();
                     tech.removeTechUnitId(id);
-                } else if (!tech.getId().equals(u.getTechId())) {
+                } else if (u.getTech() != null && !tech.getId().equals(u.getTech().getId())) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/io/CampaignXmlParser.java:739]:
```suggestion
                } else if (!tech.equals(u.getTech())) {
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 2]
```diff
@@ -3747,16 +3689,16 @@ public void removeAllTechJobsFor(Person tech) {
             return;
         }
         getHangar().forEachUnit(u -> {
-            if (tech.getId().equals(u.getTechId())) {
+            if (tech == u.getTech()) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:3692]:
```suggestion
            if (tech.equals(u.getTech())) {
```

**Review by Windchild292 — COMMENTED** [review]
We've got lots of objects being compared with == instead of .equals here.

File: MekHQ/src/mekhq/campaign/io/CampaignXmlParser.java (hunk 1 above)
Comment (sixlettervariables) [reply MekHQ/src/mekhq/campaign/io/CampaignXmlParser.java:739]: These are true reference equality.

File: MekHQ/src/mekhq/campaign/Campaign.java (hunk 2 above)
Comment (sixlettervariables) [reply MekHQ/src/mekhq/campaign/Campaign.java:3692]: These are true reference equality.

**Comment by sixlettervariables** [comment]
Post these PRs which remove UUID refs in our classes, equality will be reference equality.

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 3]
```diff
@@ -4602,43 +4515,28 @@ public void cancelMothballOrActivation() {
...
+        }
+        if (navigator != null) {
+            crew.add(navigator);
         }
         if (techOfficer != null) {
-            Person p = getCampaign().getPerson(techOfficer);
-            if (null != p) {
-                crew.add(p);
-            }
+                crew.add(techOfficer);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/Unit.java:4539]: Formatting here is odd

**Review by Windchild292 — COMMENTED** [review]
I'm still worried about using == instead of .equals because I don't trust it not causing issues in the future from a well-meaning coder.

There are some other comments too

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 4]
```diff
@@ -4326,69 +4261,51 @@ public void setScenarioId(int i) {
...
 
     public Person getTech() {
         if (null != engineer) {
             return engineer;
         }
-        if (null != tech) {
-            return getCampaign().getPerson(tech);
-        }
-        return null;
+        return tech;
```
[the remark this answers is not in the document]
Comment (sixlettervariables) [reply MekHQ/src/mekhq/campaign/unit/Unit.java:4308]: Sure, just trying for the smallest footprint diff.

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 5]
```diff
@@ -4093,10 +4093,12 @@ public void addDriver(Person p, boolean useTransfers) {
     }
 
     public void addGunner(Person p) {
-        addGunner(p, false);
+        addGunner(Objects.requireNonNull(p), false);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/Unit.java:4096]: Doesn't this duplicate the nonNull check?

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 6]
```diff
@@ -4076,10 +4074,12 @@ public boolean usesSoldiers() {
     }
 
     public void addDriver(Person p) {
-        addDriver(p, false);
+        addDriver(Objects.requireNonNull(p), false);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/Unit.java:4077]: Doesn't this duplicate the nonNull check?

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 7]
```diff
@@ -4110,10 +4112,12 @@ public void addGunner(Person p, boolean useTransfers) {
     }
 
     public void addVesselCrew(Person p) {
-        addVesselCrew(p, false);
+        addVesselCrew(Objects.requireNonNull(p), false);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/Unit.java:4115]: Doesn't this duplicate the nonNull check?

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 8]
```diff
@@ -4127,10 +4131,12 @@ public void addVesselCrew(Person p, boolean useTransfers) {
     }
 
     public void setNavigator(Person p) {
-        setNavigator(p, false);
+        setNavigator(Objects.requireNonNull(p), false);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/Unit.java:4134]: Doesn't this duplicate the nonNull check?

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 9]
```diff
@@ -4143,15 +4149,17 @@ public void setNavigator(Person p, boolean useTransfers) {
...
 
-    public boolean isTechOfficer(Person p) {
-        return Objects.equals(techOfficer, p);
+    public boolean isTechOfficer(@Nullable Person p) {
+        return (techOfficer != null) && techOfficer.equals(p);
     }
 
     public void setTechOfficer(Person p) {
-        setTechOfficer(p, false);
+        setTechOfficer(Objects.requireNonNull(p), false);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/Unit.java:4157]: Doesn't this duplicate the nonNull check?

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 10]
```diff
@@ -4193,10 +4203,12 @@ private void ensurePersonIsRegistered(Person p) {
     }
 
     public void addPilotOrSoldier(Person p) {
-        addPilotOrSoldier(p, false);
+        addPilotOrSoldier(Objects.requireNonNull(p), false);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/Unit.java:4206]: Doesn't this duplicate the nonNull check?

**Review by Windchild292 — COMMENTED** [review]
Just a last few comments

File: MekHQ/src/mekhq/campaign/unit/Unit.java (hunk 5 above)
Comment (sixlettervariables) [reply MekHQ/src/mekhq/campaign/unit/Unit.java]: Personal preference for error call stacks, but I'll simplify.

**Review by Windchild292 — APPROVED** [review]

**Comment by sixlettervariables** [comment]
@Windchild292 I went ahead and added a number of unit tests for the main methods impacted by this PR, pending CI which will run the tests I'll merge.

**Pull request merged**